Content.Server/Radio/EntitySystems/MessagesServerSystem.cs

```diff
@@ -0,0 +1,166 @@
+            }
+            _messagesCartridgeSystem.ForceUpdate(cartUid, cartComponent);
+        }
+
+        //If any names were changed or added, the server updates all the carts on its map.
+        if (toUpdate.Count > 0)
+        {
+            query = _entManager.AllEntityQueryEnumerator<MessagesCartridgeComponent>();
+            while (query.MoveNext(out var cartUid, out var cartComponent))
+            {
+                if (Transform(cartUid).MapID != mapId)
+                    continue;
+                if ((cartComponent.UserUid == null || cartComponent.UserName == null) && !(_messagesCartridgeSystem.UpdateName(cartUid, cartComponent)))
+                    continue;
+                if (cartComponent.EncryptionKey != component.EncryptionKey)
+                    continue;
+
+                foreach (var (key, value) in toUpdate)
+                {
+                    cartComponent.NameDict[key] = value;
+                }
+                _messagesCartridgeSystem.ForceUpdate(cartUid, cartComponent);
+            }
+        }
+    }
+
+    //Sync function that updates the name dictionaries of all carts to match the server.
+    //Called periodically with update.
+    public void Sync(EntityUid uid, MessagesServerComponent component)
+    {
+        var mapId = Transform(uid).MapID;
+        var query = _entManager.AllEntityQueryEnumerator<MessagesCartridgeComponent>();
+        while (query.MoveNext(out var cartUid, out var cartComponent))
+        {
+            if (Transform(cartUid).MapID != mapId)
+                continue;
+            if (cartComponent.UserUid == null || cartComponent.UserName == null)
+                continue;
+            if (cartComponent.EncryptionKey != component.EncryptionKey)
+                continue;
+
+            foreach (var entry in component.NameDict)
+            {
+                if (cartComponent.NameDict.ContainsKey(entry.Key) || cartComponent.NameDict[entry.Key] != entry.Value)
+                {
+                    cartComponent.NameDict[entry.Key] = entry.Value;
+                }
+            }
+        }
+    }
+
+    //function that tries to send a message to any matching cartridges on its map
+    public bool TryToSend(MessagesMessageData message, MapId mapId)
+    {
+        bool sent = false;
+
+        var query = EntityQueryEnumerator<MessagesCartridgeComponent, CartridgeComponent>();
+
+        while (query.MoveNext(out var uid, out var messagesCartridgeComponent, out var cartridge))
+        {
+            if (Transform(uid).MapID != mapId)
+                continue;
+            if (cartridge.LoaderUid != null) //<TODO> this should probably be more generalisable
+            {
+                if (messagesCartridgeComponent.UserUid == message.ReceiverId)
+                    _messagesCartridgeSystem.ServerToPdaMessage(uid, messagesCartridgeComponent, message, cartridge.LoaderUid.Value);
+                sent = true;
+            }
+        }
+
+        return sent;
+    }
```

Comment:
Okay having read this after the above you should just be caching what message cartridges are actually relevant for a server up-front, then re-using that data throughout the rest of the update. As it is if we have 10 messages to send you iterate every single message cartridge 10 times instead of the once per relevant cartridge.